Firestore/core/test/firebase/firestore/remote/grpc_connection_test.mm

```diff
@@ -0,0 +1,169 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <memory>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "Firestore/core/src/firebase/firestore/auth/token.h"
+#include "Firestore/core/src/firebase/firestore/remote/connectivity_monitor.h"
+#include "Firestore/core/src/firebase/firestore/remote/grpc_connection.h"
+#include "Firestore/core/src/firebase/firestore/util/async_queue.h"
+#include "Firestore/core/src/firebase/firestore/util/executor_std.h"
+#include "Firestore/core/test/firebase/firestore/util/grpc_stream_tester.h"
+#include "absl/memory/memory.h"
+#include "gtest/gtest.h"
+
+namespace firebase {
+namespace firestore {
+namespace remote {
+
+using auth::Token;
+using auth::User;
+using core::DatabaseInfo;
+using model::DatabaseId;
+using util::AsyncQueue;
+using util::GrpcStreamTester;
+using util::Status;
+using util::StatusOr;
+using util::internal::ExecutorStd;
+using NetworkStatus = ConnectivityMonitor::NetworkStatus;
+
+namespace {
+
+class MockConnectivityMonitor : public ConnectivityMonitor {
+ public:
+  explicit MockConnectivityMonitor(AsyncQueue* worker_queue)
+      : ConnectivityMonitor{worker_queue} {
+    SetInitialStatus(NetworkStatus::Available);
+  }
+
+  void set_status(NetworkStatus new_status) {
+    MaybeInvokeCallbacks(new_status);
+  }
+};
+
+bool IsConnectivityChange(const Status& status) {
+  return status.code() == FirestoreErrorCode::Unavailable &&
+         status.error_message() == "Network connectivity changed";
+}
+
+class ConnectivityObserver : public GrpcStreamObserver {
+ public:
+  void OnStreamStart() override {
+  }
+  void OnStreamRead(const grpc::ByteBuffer& message) override {
+  }
+  void OnStreamFinish(const util::Status& status) override {
+    if (IsConnectivityChange(status)) {
+      ++connectivity_change_count_;
+    }
+  }
+
+  int connectivity_change_count() const {
+    return connectivity_change_count_;
+  }
+  int connectivity_change_count_ = 0;
+};
+
+}  // namespace
+
+class GrpcConnectionTest : public testing::Test {
+ public:
+  GrpcConnectionTest() {
+    auto connectivity_monitor_owning =
+        absl::make_unique<MockConnectivityMonitor>(&tester->worker_queue());
+    connectivity_monitor = connectivity_monitor_owning.get();
+    tester = absl::make_unique<GrpcStreamTester>(
+        std::move(connectivity_monitor_owning));
+  }
+
+  void SetNetworkStatus(NetworkStatus new_status) {
+    tester->worker_queue().EnqueueBlocking(
+        [&] { connectivity_monitor->set_status(new_status); });
+    // Make sure the callback executes.
+    tester->worker_queue().EnqueueBlocking([] {});
+  }
+
+  std::unique_ptr<GrpcStreamTester> tester;
+  MockConnectivityMonitor* connectivity_monitor = nullptr;
+};
+
+TEST_F(GrpcConnectionTest, GrpcStreamsNoticeChangeInConnectivity) {
+  ConnectivityObserver observer;
+
+  auto stream = tester->CreateStream(&observer);
```

Comment:
My preference would be to use the type here, rather than auto. Both here and throughout. Optional.

Reply:
Done.

I'm slightly inclined towards auto when the function name pretty much contains the type name (also the type is a little unwieldy, due to `unique_ptr`), but I don't feel strongly about it.